### DISCUSSION TITLE
cancellation token in orchestrators

### DISCUSSION THREAD

**Discussion by ghost** [Azure/azure-functions-durable-extension#1832] [Q&A]
I have an infinite loop kind of situation for orchestrators(healthy loop). With the help of timer job we create an orchestrator and it triggers an activity. after the end of orchestrator, it triggers another orchestrator. This way our orchestrator keeps running until the job is done. 
I want a graceful shutdown of host. i mean my orchestrator should not fire a new orchestrator and let the running activity end. But how can i get the cancellation token parameter into orchestrator, without the timer job being fired.

**Comment by cgillum** [comment]
Can you share a simple example of what your orchestration code looks like? It doesn't normally make sense to use a cancellation token inside of an orchestrator token, but maybe there is another solution we could explore.

**Reply by ghost** [reply]
```csharp
[FunctionName("myorch")]
public async Task RunOrchestratorAsync(
    [OrchestrationTrigger] IDurableOrchestrationContext context)
{
    Input inputToken = context.GetInput<Input>();

    Input activityOutput = await context.CallActivityAsync<Input>(
        "myactivity",
        inputToken);

    if (activityOutput.ShouldContinue)
    {
        // we keep calling activity [indirectly] till all the data is processed
        ActivityInput newOrchestratorInput = new ActivityInput();

        context.StartNewOrchestration(
            "myorch",
            newOrchestratorInput,
            new Guid());
    }
    else
    {
        //stop orchestartor.
    }
}
```

**Reply by ghost** [reply]
the problem is if the activity was running and we shutdown the host. will the activity finish of and trigger another orchestrator? I think it is guaranteed that the in process activity goes to completion(my activity is idempotent only if it goes to completion)

**Reply by ghost** [reply]
@cgillum is it guaranteed that if activity starts will it goes to completion, even if there is a manual restart of host?

**Reply by ghost** [reply]
@cgillum can you help me with this. will the running activity go to to completion(my activity runs not more than 1 minute) if there is a manual restart of the host from portal.

**Comment by cgillum** [comment]
> is it guaranteed that if activity starts will it goes to completion, even if there is a manual restart of host?

Yes, Durable Functions guarantees that the activity will _eventually_ finish. However, we can't guarantee that it will finish before the process gets recycled if you click the restart button. If the activity execution gets interrupted by a host recycle, it will run again after about 5 minutes. This means that it's possible that the activity will run twice if the first execution gets interrupted before its completion can be recorded.
